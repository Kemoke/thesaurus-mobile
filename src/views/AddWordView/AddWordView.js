import React from 'react';
import styles from './style';
import {
    ScrollView,
    Text,
    TouchableOpacity,
    View
} from "react-native";
import {Button, Header, Input} from 'react-native-elements';
import {Ionicons} from "@expo/vector-icons";
import {SynonymContextWrapper} from "../../SynonymContext";
import type {SynonymContextProps} from "../../SynonymContext";
import KeyboardAvoidingView from "../../util/KeyboardAvoidingView";


type AddWordViewProps = {
    synonymContext: SynonymContextProps
}

const initState = {
    word: '',
    synonyms: ['']
};

export class AddWordViewComponent extends React.Component<AddWordViewProps> {
    state = initState;

    onSave = () => {
        const {word, synonyms} = this.state;
        synonyms.forEach(synonym => {
            this.props.synonymContext.addSynonym(word, synonym)
        });
        this.setState(initState);
    };

    onSynonymChange = (synonym, index) => {
        const synonyms = [...this.state.synonyms];
        synonyms[index] = synonym;
        this.setState({synonyms});
    };

    addSynonym = () => this.setState({synonyms: [...this.state.synonyms, '']});

    removeSynonym = (index) => {
        const synonyms = [...this.state.synonyms];
        synonyms.splice(index, 1);
        this.setState({synonyms});
    };

    render() {
        const {word, synonyms} = this.state;
        return (
            <KeyboardAvoidingView verticalOffset={40}>
                <Header centerComponent={{text: 'ADD NEW WORD', style: {color: "#FFF"}}}/>
                <ScrollView contentContainerStyle={styles.mainContainer}>
                    <Text style={styles.label}>Word</Text>
                    <Input placeholder="Enter word"
                           value={word}
                           testID='wordInput'
                           onChangeText={word => this.setState({word})}/>
                    <Text style={styles.label}>Synonyms</Text>
                    {synonyms.map((synonym, index) => (
                        <View key={index} style={styles.synonymInputContainer}>
                            <Input
                                placeholder="Enter synonym"
                                containerStyle={styles.synonymInput}
                                value={synonym}
                                testID={`synonymInput${index}`}
                                onChangeText={text => this.onSynonymChange(text, index)}/>
                            <Ionicons name='md-trash'
                                      size={25}
                                      color='#dc3545'
                                      testID={`synonymRemove${index}`}
                                      style={styles.deleteIcon}
                                      onPress={() => this.removeSynonym(index)}/>
                        </View>
                    ))}
                    <TouchableOpacity style={styles.addButton}
                                      testID={`addSynonym`}
                                      onPress={this.addSynonym}>
                        <Ionicons name={'ios-add'}
                                  size={25}
                                  style={styles.addIcon}
                                  color='#3CBD3F'/>
                        <Text style={styles.addText}>Add synonym</Text>
                    </TouchableOpacity>
                    <View style={styles.saveButtonContainer}>
                        <View style={styles.saveButton}>
                            <Button title="Save"
                                    onPress={this.onSave}/>
                        </View>
                    </View>
                </ScrollView>
            </KeyboardAvoidingView>
        );
    }
}

export default SynonymContextWrapper(AddWordViewComponent)