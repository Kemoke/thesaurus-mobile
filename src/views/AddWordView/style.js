import {StyleSheet} from "react-native";

export default StyleSheet.create({
    mainContainer: {
        paddingTop: 16,
        paddingLeft: 16,
        paddingRight: 16,
        paddingBottom: 32,
        backgroundColor: '#FFF'
    },
    synonymInputContainer: {
        flexDirection: 'row',
        alignItems: 'flex-end',
        justifyContent: 'space-between',
        width: '100%',
        marginBottom: 8
    },
    synonymInput: {
        width: '90%'
    },
    label: {
        fontSize: 16,
        marginLeft: 10,
        marginTop: 16,
    },
    deleteIcon: {
        marginRight: 12
    },
    addButton: {
        flexDirection: 'row',
        alignItems: 'center',
        marginLeft: 10,
        marginTop: 8,
        width: 150
    },
    addText: {
        color: '#3CBD3F'
    },
    addIcon: {
        marginRight: 8
    },
    saveButtonContainer: {
        alignItems: 'center',
        marginTop: 16
    },
    saveButton: {
        width: 150,
    }
})