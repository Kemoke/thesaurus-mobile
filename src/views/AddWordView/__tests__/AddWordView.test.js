import React from "react";
import {shallow} from 'enzyme'
import {AddWordViewComponent} from "../AddWordView";

describe('Add Word View', () => {
    it('should render correctly', function () {
        const component = shallow(
            <AddWordViewComponent synonymContext={{}}/>
        );
        expect(component).toMatchSnapshot();
    });
    it('should handle saving synonyms', function () {
        const context = {
            addSynonym: jest.fn(),
            words: {}
        };
        const component = shallow(
            <AddWordViewComponent synonymContext={context}/>
        );
        component.setState({word: 'test', synonyms: ['test1', 'test2']});
        component.instance().onSave();
        expect(context.addSynonym).toHaveBeenCalledWith('test', 'test1');
        expect(context.addSynonym).toHaveBeenCalledWith('test', 'test2');
        expect(component.state()).toEqual({word: '', synonyms: ['']})
    });
    it('should handle word input', function () {
        const component = shallow(
            <AddWordViewComponent synonymContext={{}}/>
        );
        component.findWhere(node => node.prop('testID') === 'wordInput').props().onChangeText("test");
        expect(component.state("word")).toEqual("test")
    });
    it('should handle synonym input', function () {
        const component = shallow(
            <AddWordViewComponent synonymContext={{}}/>
        );
        component.setState({synonyms: ['test1', 'test2', 'test3']});
        component.findWhere(node => node.prop('testID') === 'synonymInput0').props().onChangeText("testnew1");
        expect(component.state("synonyms")[0]).toEqual("testnew1");
        component.findWhere(node => node.prop('testID') === 'synonymInput1').props().onChangeText("testnew2");
        expect(component.state("synonyms")[1]).toEqual("testnew2");
        component.findWhere(node => node.prop('testID') === 'synonymInput2').props().onChangeText("testnew3");
        expect(component.state("synonyms")[2]).toEqual("testnew3");
    });
    it('should handle adding synonyms', function () {
        const component = shallow(
            <AddWordViewComponent synonymContext={{}}/>
        );
        component.findWhere(node => node.prop('testID') === 'addSynonym').props().onPress();
        expect(component.state("synonyms").length).toEqual(2)
    });
    it('should handle removing synonyms', function () {
        const component = shallow(
            <AddWordViewComponent synonymContext={{}}/>
        );
        component.setState({synonyms: ['test1', 'test2', 'test3']});
        component.findWhere(node => node.prop('testID') === 'synonymRemove1').props().onPress();
        expect(component.state("synonyms")).toEqual(["test1", "test3"]);
    });
});