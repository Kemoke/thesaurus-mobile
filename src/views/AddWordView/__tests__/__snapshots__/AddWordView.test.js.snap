// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Word View should render correctly 1`] = `
<KeyboardAvoidingView
  verticalOffset={40}
>
  <Themed.Header
    centerComponent={
      Object {
        "style": Object {
          "color": "#FFF",
        },
        "text": "ADD NEW WORD",
      }
    }
  />
  <ScrollViewMock
    contentContainerStyle={
      Object {
        "backgroundColor": "#FFF",
        "paddingBottom": 32,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      }
    }
  >
    <Component
      style={
        Object {
          "fontSize": 16,
          "marginLeft": 10,
          "marginTop": 16,
        }
      }
    >
      Word
    </Component>
    <ForwardRef(Themed.Input)
      onChangeText={[Function]}
      placeholder="Enter word"
      testID="wordInput"
      value=""
    />
    <Component
      style={
        Object {
          "fontSize": 16,
          "marginLeft": 10,
          "marginTop": 16,
        }
      }
    >
      Synonyms
    </Component>
    <Component
      key="0"
      style={
        Object {
          "alignItems": "flex-end",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 8,
          "width": "100%",
        }
      }
    >
      <ForwardRef(Themed.Input)
        containerStyle={
          Object {
            "width": "90%",
          }
        }
        onChangeText={[Function]}
        placeholder="Enter synonym"
        testID="synonymInput0"
        value=""
      />
      <Icon
        allowFontScaling={false}
        color="#dc3545"
        name="md-trash"
        onPress={[Function]}
        size={25}
        style={
          Object {
            "marginRight": 12,
          }
        }
        testID="synonymRemove0"
      />
    </Component>
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginLeft": 10,
          "marginTop": 8,
          "width": 150,
        }
      }
      testID="addSynonym"
    >
      <Icon
        allowFontScaling={false}
        color="#3CBD3F"
        name="ios-add"
        size={25}
        style={
          Object {
            "marginRight": 8,
          }
        }
      />
      <Component
        style={
          Object {
            "color": "#3CBD3F",
          }
        }
      >
        Add synonym
      </Component>
    </TouchableOpacity>
    <Component
      style={
        Object {
          "alignItems": "center",
          "marginTop": 16,
        }
      }
    >
      <Component
        style={
          Object {
            "width": 150,
          }
        }
      >
        <ForwardRef(Themed.Button)
          onPress={[Function]}
          title="Save"
        />
      </Component>
    </Component>
  </ScrollViewMock>
</KeyboardAvoidingView>
`;
